crates/adapters/src/format/avro/schema.rs

```diff
@@ -0,0 +1,238 @@
+//! Helpers for working with Avro schemas.
+
+use apache_avro::{schema::RecordField, Schema as AvroSchema};
+use feldera_types::program_schema::{canonical_identifier, ColumnType, Field, SqlType};
+
+/// Convert schema to JSON format.
+pub fn schema_json(schema: &AvroSchema) -> String {
+    serde_json::to_string(schema).unwrap_or_else(
+        // This should never happen, but just in case.
+        |_| "Avro schema cannot be converted to JSON".to_string(),
+    )
+}
+
+/// Extract value type schema from a nullable field schema, which has the following shape:
+/// Union[null, type] or Union[type, null]. Returns `None` if `schema` doesn't match this
+/// pattern.
+pub fn nullable_schema_value_schema(schema: &AvroSchema) -> Option<&AvroSchema> {
+    match schema {
+        AvroSchema::Union(union_schema) => match union_schema.variants() {
+            [AvroSchema::Null, s] => Some(s),
+            [s, AvroSchema::Null] => Some(s),
+            _ => None,
+        },
+        _ => None,
+    }
+}
+
+/// Find a field in a record schema.
+fn lookup_field<'a>(fields: &'a [RecordField], field: &'a Field) -> Option<&'a RecordField> {
+    let name = field.name.name();
+
+    // TODO: check `record_field.aliases`.
+    fields
+        .iter()
+        .find(|&record_field| canonical_identifier(&record_field.name) == name)
```

Comment:
I think you probably want SqlIdentifier in RecordField.name

Reply:
we should make canonical_identifier private (I thought I did)

Reply:
This one I don't control